subprojects/logging/src/main/java/org/gradle/internal/logging/sink/GroupingProgressLogEventGenerator.java

```diff
@@ -68,6 +83,10 @@ public void onOutput(OutputEvent event) {
         } else if (event instanceof ProgressCompleteEvent) {
             onComplete((ProgressCompleteEvent) event);
         } else if (event instanceof EndOutputEvent) {
+            if (future != null && !future.isCancelled()) {
+                future.cancel(false);
+            }
+            executor.shutdown();
```

Comment:
If I remember correctly, it's safe to assume the `EndOutputEvent` is going to be sent for every build.